discovery/dockerswarm/tasks.go

```diff
@@ -0,0 +1,164 @@
+			lbs, ports, err := d.getServiceLabelsAndPorts(ctx, s.ServiceID)
+			if err != nil {
+				return nil, fmt.Errorf("error while inspecting service %s: %w", s.ServiceID, err)
+			}
+			serviceLabels[s.ServiceID] = lbs
+			servicePorts[s.ServiceID] = ports
+		}
+
+		commonLabels := model.LabelSet{
+			model.LabelName(swarmLabelTaskID):           model.LabelValue(s.ID),
+			model.LabelName(swarmLabelTaskDesiredState): model.LabelValue(s.DesiredState),
+			model.LabelName(swarmLabelTaskStatus):       model.LabelValue(s.Status.State),
+			model.LabelName(swarmLabelTaskSlot):         model.LabelValue(fmt.Sprintf("%v", s.Slot)),
+		}
+
+		if s.Status.ContainerStatus != nil {
+			commonLabels[model.LabelName(swarmLabelTaskContainerID)] = model.LabelValue(s.Status.ContainerStatus.ContainerID)
+		}
+
+		for k, v := range s.Labels {
+			ln := strutil.SanitizeLabelName(k)
+			commonLabels[model.LabelName(swarmLabelTaskLabelPrefix+ln)] = model.LabelValue(v)
+		}
+
+		for k, v := range serviceLabels[s.ServiceID] {
+			commonLabels[k] = v
+		}
+
+		for k, v := range nodeLabels[s.NodeID] {
+			commonLabels[k] = v
+		}
+
+		for _, p := range s.Status.PortStatus.Ports {
+			if p.Protocol != swarm.PortConfigProtocolTCP {
+				continue
+			}
+
+			labels := model.LabelSet{
+				model.LabelName(swarmLabelTaskPortMode): model.LabelValue(p.PublishMode),
+			}
+
+			for k, v := range commonLabels {
+				labels[k] = v
+			}
+
+			for k, v := range serviceLabels[s.ServiceID] {
+				labels[k] = v
+			}
+
+			for k, v := range nodeLabels[s.NodeID] {
+				labels[k] = v
+			}
+
+			addr := net.JoinHostPort(string(labels[swarmLabelNodeAddress]), strconv.FormatUint(uint64(p.PublishedPort), 10))
+			labels[model.AddressLabel] = model.LabelValue(addr)
+			tg.Targets = append(tg.Targets, labels)
+			continue
+		}
+
+		for _, p := range servicePorts[s.ServiceID] {
+			if p.Protocol != swarm.PortConfigProtocolTCP {
+				continue
+			}
+			for _, network := range s.NetworksAttachments {
+				for _, address := range network.Addresses {
+					labels := model.LabelSet{
+						model.LabelName(swarmLabelTaskPortMode): model.LabelValue(p.PublishMode),
+					}
+
+					for k, v := range commonLabels {
+						labels[k] = v
+					}
+
+					if _, ok := networkLabels[network.Network.ID]; !ok {
+						lbs, err := d.getNetworkLabels(ctx, network.Network.ID)
+						if err != nil {
+							return nil, fmt.Errorf("error while inspecting network %s: %w", network.Network.ID, err)
+						}
+						networkLabels[network.Network.ID] = lbs
+					}
+
+					for k, v := range networkLabels[network.Network.ID] {
+						labels[k] = v
+					}
+
+					ip, _, err := net.ParseCIDR(address)
+					if err != nil {
+						return nil, fmt.Errorf("error while parsing address %s: %w", address, err)
+					}
+					addr := net.JoinHostPort(ip.String(), strconv.FormatUint(uint64(p.PublishedPort), 10))
+					labels[model.AddressLabel] = model.LabelValue(addr)
+
+					tg.Targets = append(tg.Targets, labels)
```

Comment:
Why are we adding two different types of target?

Reply:
Because there are two ways exposing tasks

https://docs.docker.com/engine/swarm/services/#publish-ports

- Publish a service’s ports using the routing mesh
- Publish a service’s ports directly on the swarm node

Reply:
Should they be different roles?

Reply:
No because from a users and commands perspective they are really similar.

Reply:
We should probably clarify a little in the docs which labels are present for each then.

Reply:
It is here already:

The __meta_dockerswarm_network_* meta labels are not populated for ports which are published with mode=host.